### DISCUSSION TITLE
Half-rotations increasing without water flow (Badger Model 25 Recordall)

### DISCUSSION THREAD

**Discussion by eigenphase** [tronikos/esphome-magnetometer-water-gas-meter#33] [Q&A]
So awesome project! My meter is outside on the street, too little clearance for proximity sensor or AI-on-the-edge so I didn't expect I could ever read it but thanks to this project I am very close :-) 

I have a Badger Model 25 Recordall and I am observing that the number of half rotations continuously increases without water flow:

<img width="877" alt="image" src="https://github.com/user-attachments/assets/6c96b4b9-54da-49b5-8525-8afb7351bfab">

I installed the sensor at ~11:52. You can clearly see the two times I turned on the water flow. During these times, all three axis look sinusoidal as expected. But when there is no water flow, the magnetic sensors still change, causing the number of half rotations to rise.

Is this normal? I suspect this is just ambient noise (maybe cars driving by? Slight vibrations?) Is there anything that can be done about it?

I believe I have found the right manual here: https://www.usabluebook.com/product/attachment/54253/542523.pdf . On PDF page 16 it states for M25 (5/8") 198.334 pulses/gallon (unscaled). From the plot above I estimate 31 half rotations per  10 minutes, so ~16 rotations per 10 minutes or 1.6 rotations per minute. So 1.6/198.334=0.008067199774 gallons per minute. Or, multiplying with 60*24*31, that would be ~360 gallons per month. Sounds extremely high to me.

**Comment by tronikos** [comment]
This is not normal. I get fairly flat lines when water isn't running.

<img width="1714" alt="image" src="https://github.com/user-attachments/assets/c38ad784-84f0-458e-b411-5b7fa1d3d9ab">

Make sure it's well mounted and doesn't move and that you don't have a small leak that you aren't aware of. I'd try moving it far from the meter and keep it stable to see if you get flat lines. You might also want to try the Sensors app on your phone. My guess is you got a bad magnetometer and you need to replace it.

**Reply by eigenphase** [reply]
Thank you very much!! I have two sensors but I did some debugging and it's the water meter.

I stuck the sensor on the wall for a couple of hours and the magnetic field was pretty much constant.

Then I put my phone next to the meter and I can see pretty much the same changes as the sensor gets (of course, shape and direction is different ... but it's always flat for some time and then there's a change):

<img width="426" alt="image" src="https://github.com/user-attachments/assets/a36eed25-a0ec-4204-a0d9-d97ab083cf0d">

Lastly, I put the sensor back to the meter (for now, I fixed it with double sided tape only but should be good enough), this time the other side. But the behavior is exactly the same:

<img width="584" alt="image" src="https://github.com/user-attachments/assets/f0d30da0-205c-4b47-b3d2-2bebf10a504c">


Now, if there would be a leak, shouldn't we see a sinusoid of a very low frequency? What we see are small durations of flat and then sudden changes, up and down.

Finally I recorded a video of the flow meter. The video is too large to attach it here put here: https://streamable.com/uxg79p


Now this is super interesting: Indeed the flow meter (the blue dial) is indeed moving, but it's moving back and forth. It feels as if the disc would just be "wobbling around". Now that the blue dial is not just moving forward, but forth **and back**, I assume that the average will be zero.

1. Any plausible explanation for this? Possibly not much pressure in the line and the disk indeed just wobbles back and forth even if there's no flow? 
2. Is there a way to add a filter to the raw data? It seems that in the current code, the rotation is strictly monotonic, i.e., it only increases. But I assume this is wrong: There should be a movement of the disk and in one direction it goes up, in the other direction it goes down.

**Reply by tronikos** [reply]
When water is running, is the blue dial rotating in one direction? Or does it similarly go back and forth like in the video? If it also goes back and forth when water is running you might have a small leak.
Otherwise it seems you don't have a closed water system. You need check valves, backflow prevention devices, pressure-reducing valves, or other devices on the supply line to prevent water from being sucked back into the municipal water supply.
Unfortunately there is no easy way to detect the direction of the flow in software.

**Reply by eigenphase** [reply]
1.) Yes, when the water runs the flow indicator (blue dial) only goes one direction

2.) Ok, I did a bit of research and it there are many people having the same issue and the conclusion is always the same: This is normal and can happen due to pressure changes on either side of the meter.

I have also seen people report it being due to their water heater. I turned off the valve to my water heater and lo and behold: It stayed constant! So I think what happens is that some hot water from my water heater leaks through the mixing valve and creates pressure differences. This causes a slight back-and-forth flow between the utility line and the house line.

Yes, you are right, a check valve or backflow preventer would avoid this. I called my utility and unfortunately they do NOT install any backflow prevention at the meter. I asked them if they would or if they would if I purchased the valve but they refuse. Also I am not allowed to install such a valve myself in the meter box. They said a backflow preventing device is not required to be installed for new construction and they also hinted they may require the customer to install one even for old construction (like mine). While I agree there should be one and I'd really like to have one, it seems this would be entirely on me and cost hundreds if not thousands of dollars. So not an option.

But how is Flume doing it then? They also just measure the magnetic field. I am sure there are very many old homes without any check valves/backflow prevention, so this can't be too uncommon. I did not find any indication that Flume would have an issue (even though I'm not able to test it myself since I don't have one).

I agree there is no easy way to detect the direction of the flow in SW because the waveform H(t) can go up and down even for desired flow. However, can you still think of a conceptual way to filter such data out? I am trying to understand this better myself and I may give it a stab.

**Reply by tronikos** [reply]
Do you have a water heater expansion tank? Adding one or removing it might help.
I have an open water system, like yours, without a water heater expansion tank, and I haven't noticed your issue.
No idea what Flume is doing.

**Reply by tech-rat** [reply]
I have been working on this same issue. I have an expansion tank AND check valve but the amount of water that goes back and forth is very small. But when the meter sensitivity is hundredths of a gallon it still looks like water usage all the time.

![image](https://github.com/user-attachments/assets/0e6b377d-ce75-4733-8796-55035f95f6d6)

So I built a DUAL sensor model. I'm not fluent in the ESPHome/YAML code so I write my prototype in C/Wiring.
The logic reads the Z-axis on both sensors and converts it to binary.
Then using the two signals treats it like a rotary encoder.
That way the pulse counter goes up and down and my reading seems accurate.

I have only gotten far enough to record the number of pulses and the actual reading from the meter but my graph looks straight so I think I'm onto something.

![image](https://github.com/user-attachments/assets/7d75e509-6f4a-4729-bdb7-78c7bbc41746)

**Reply by eigenphase** [reply]
First are you sure you really have a check valve? That should really fix it.

I actually got a Sharkbite check valve and managed to put it into my main water line within 30min. This fixed the issue for me. Actually I’m not at home right now and it shows 0.07gal/day. Not sure if normal leakage or measured error/noise.

having said that, really impressive your idea. I don’t have much experience in esphome either but I’ve played around a bit. Maybe you can share your code?

We may need to add C code directly, currently this project works with only built in code. Together we should be able to make it work. Maybe even @tronikos would be willing to help?

**Reply by dnschneid** [reply]
I used the quadrature encoder method in my [espleak](https://github.com/dnschneid/espleak/blob/main/espleak.yaml) project (inspired by this one); maybe you can crib something from that.

**Reply by eigenphase** [reply]
Absolutely amazing! I ended up installing a check valve and for me it’s good now.

**Reply by tech-rat** [reply]
Awesome. Will check it out. Thanks

**Reply by tronikos** [reply]
@dnschneid Nice work! Clever using 2 magnetometers. While you are here, could you help with #34 that is also using 2 magnetometers?

**Reply by jsermer** [reply]
I was initially having this same behavior until I adjusted the calibration time to be 15s and pressed 'calibrate axis'.  When I initially calibrated it with the default of 5s, it adjusted itself to the x axis, 4.8uT lower threshold, and 14.8uT upper threshold.  With a longer calibration time, it adjusted itself to the z axis, 18.4uT lower threshold, and 28.4uT upper threshold.  I manually set the meter total to match the current meter reading, and after almost a day of use, the home assistant sensor nearly matches the meter.  FWIW, I taped the sensor laying flat against the front of the meter like the pictures in the guide show.

**Reply by eigenphase** [reply]
@dnschneid I mentioned above the check valve fixed the issue but it seems not perfectly. I still have a few very low (0.06gal/min) pulses spread throughout the way. When not at home, it's impossible to find them. But I wasn't at home for multiple days and I counted up to 180 such pulses.

So I'd like to consider your quadrature encoding method. Do you have any reference on how it works? I'd really like to understand it. Also, why are really two devices needed? Why can't we just use two directions (e.g. X and Z) of one sensor? And for my meter (the Badger Recordall 25) I mounted the sensor already to the best possible place. If I put another on the 90 degree side, this signal will be weaker. Is this an issue for your method?

**Reply by dnschneid** [reply]
The current sensing method (quadrature or not) takes the raw magnetic data and converts it to binary (positive or negative), which is resilient to noise and sample timing.  Quadrature encoding uses two binary sensors at a phase offset to detect rotation/speed; if we're using the same sensing method then two physical sensors placed at a phase offset are required.

If you don't have room on your sensor (badger 25 should have plenty of room though), they don't have to be perfectly 90 degrees relative to each other, but the further from 90 degrees, the lower the maximum flow rate you can reliably detect.

I think in theory we could use X (or Y, depending on how your sensor is oriented) as a way to disambiguate the direction of rotation, except the signal is pretty weak and we'd have to sample fast enough so we don't get confused and think it's the opposite direction.  I suspect the maximum flow rate would be severely limited by this, unless we did something smart and only used the X reading to determine direction at low flow rates.  Not really sure how that algorithm would look, but it seems plausible if someone wanted to experiment.

**Reply by eigenphase** [reply]
Thanks! Btw, I found now a reference in your code: https://cdn.sparkfun.com/datasheets/Robotics/How%20to%20use%20a%20quadrature%20encoder.pdf

Is there any disadvantage of using espleak instead of this project? I do have additional magnetometers, so I may make this my next weekend project.

Initially I thought I'd need multiple watermeters ... but mounting two magnetometers should be doable.

**Reply by dnschneid** [reply]
Needing two sensors attached is a pretty big disadvantage.  Harder to set up, needs more wires, works the esp32 harder, increased risk of sensor failure.

If you're OK with the increased setup complexity and the lack of gas meter support, espleak gains you reverse flow detection and leak response.  If you don't need that (not you specifically, segdy) then better to stick with troniko's single-sensor solution.
